backend/src/test/java/ee/tlu/forum/controller/PostControllerTest.java

```diff
@@ -0,0 +1,244 @@
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        when (postService.getPostById(any()))
+                .thenReturn(p1);
+
+        // when then
+        mockMvc.perform(MockMvcRequestBuilders.get("/api/post/1"))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.status().is(200))
+                .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.id", Matchers.is(1)))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.text", Matchers.is("text")));
+    }
+
+    @Test
+    @DisplayName("Returns created post - POST /api/post/add")
+    void createPost() throws Exception {
+        // given
+        User user = new User(1L,
+                "user1",
+                "test1@test.com",
+                "aaa",
+                new ArrayList<>(),
+                "",
+                "",
+                new ArrayList<>(),
+                new ArrayList<>());
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        AddNewPostInput form = new AddNewPostInput();
+        form.setUsername("user1");
+        form.setText("text");
+        String formJson = new ObjectMapper().writeValueAsString(form);
+
+        when (postService.createPost(any(AddNewPostInput.class)))
+                .thenReturn(p1);
+
+        RequestBuilder requestBuilder = MockMvcRequestBuilders.post("/api/post/add")
+                .accept(MediaType.APPLICATION_JSON)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(formJson);
+
+        // when then
+        mockMvc.perform(requestBuilder)
+                .andExpect(MockMvcResultMatchers.status().is(201))
+                .andDo(MockMvcResultHandlers.print())
+                .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.id", Matchers.is(1)))
+                .andExpect(MockMvcResultMatchers.jsonPath("$.text", Matchers.is("text")));
+    }
+
+    @Test
+    @DisplayName("Returns edited post - PATCH /api/post/edit")
+    void editPost() throws Exception {
+        // given
+        User user = new User(1L,
+                "user1",
+                "test1@test.com",
+                "aaa",
+                new ArrayList<>(),
+                "",
+                "",
+                new ArrayList<>(),
+                new ArrayList<>());
+        Thread t1 = new Thread(1L, "content", "title", user, new ArrayList<>());
+        Post p1 = new Post(1L, "text", user, t1);
+
+        Map<String, String> form = new HashMap<>();
+        form.put("title", t1.getTitle());
+        String formJson = new ObjectMapper().writeValueAsString(form);
+
```

Comment:
I think I'm getting confused here but I don't see us actually changing anything about the post?

Reply:
We mock the actual output of the service layer - there's no repository or database in tests that would change it. This makes sure that the request was valid and actually outputs the above post, even if nothing was changed. I can indeed change the title and add a line that changes p1's title, but it would accomplish the same thing, as it's still a valid request.

I can add a todo if it's necessary?

Reply:
Ah OK, I'd probably mock out both states (original and then edited) and check that it actually changes but that's probably just me.
I shall leave it up to you to decide whether you can be bothered or not. I don't think it's a deal breaker.